confluent_kafka/avro/serializer/__init__.py

```diff
@@ -33,8 +61,270 @@ def __str__(self):
+    """
+    Constructs the subject name under which a schema is registered with the Confluent Schema Registry.
+
+    TopicNameStrategy returns the schema's subject in the form of <topic>-key or <topic>-value.
+
+    :param str topic: Topic name.
+    :param is_key: True if subject is being registered for a message key.
+    :param schema schema: Parsed Avro schema. *Note* Not used by TopicNameStrategy
+    :raises ValueError: If topic is unset.
+    :returns: The subject name with which to register the schema.
+    :rtype: str
+    """
+    if topic is None:
+        raise ValueError("Topic must be set when using TopicNameStrategy")
+
+    return "-".join([topic, '-key' if is_key else '-value'])
+
+
+def RecordNameStrategy(topic=None, is_key=False, schema=None):
+    """
+    Constructs the subject name under which a schema is registered with the Confluent Schema Registry.
+
+    RecordNameStrategy returns the fully-qualified record name regardless of the topic.
+
+    Compatibility checks of the same record name across all topics.
+    This strategy allows a topic to contain a mixture of different record types.
+
+    :param str topic: Topic name.  *Note* Not used by RecordNameStrategy
+    :param is_key: True if subject is being registered for a message key. *Note* Not used by RecordNameStrategy.
+    :param schema schema: Parsed Avro schema.
+    :raises ValueError: If schema is not set.
+    :returns: The subject name with which to register the schema.
+    :rtype: str
+    """
+    if schema is None:
+        raise ValueError("Schema must be set when using RecordNameStategy")
+
+    return schema.fullname
+
+
+def TopicRecordNameStrategy(topic=None, is_key=False, schema=None):
+    """
+    Constructs the subject name under which a schema is registered with the Confluent Schema Registry.
+
+    TopicRecordNameStrategy returns the topic name appended by the fully-qualified record name.
+
+    Compatibility checks are performed against all records of the same name within the same topic.
+    Like the RecordNameStrategy mixed record types are allowed within a topic.
+    This strategy is more flexible in that records needn't be complaint across the cluster.
+
+    :param str topic: Topic name.
+    :param schema schema: Parsed Avro schema.
+    :param is_key: True if used by a key_serializer.
+    :raises ValueError: If topic and schema are not set.
+    :returns: The subject name with which to register the schema.
+    :rtype: str
+    """
+    if not any([topic, schema]):
+        raise ValueError("Both Topic and Schema must be set when using TopicRecordNameStrategy")
+    return "-".join([topic, schema.fullname])
+
+
+class AvroSerializer(object):
+    """
+    Encodes kafka messages as Avro; registering the schema with the Confluent Schema Registry.
+
+    :param registry_client CachedSchemaRegistryClient: Instance of CachedSchemaRegistryClient.
+    :param bool is_key: True if configured as a key_serializer.
```

Comment:
It a much better abstraction imo if `is_key` is specified at serialization time, not as a constructor argument. that would also give consistency with how you've implemented `subject_strategy`.

Reply:
I disagree. Since I am explicitly setting a key or value serializer I see no need to restate that with each record I intend to serialize. Subject naming strategies are different as they are a schema registry client construct and not directly related to serialization.

Reply:
why is needing to restate this with each record serialized a disadvantage? it's a trivial internal detail. by contrast, having serializer instances tied keys or values results in an externally exposed abstraction that is more complex. strongly disagree with your disagree.

Reply:
But I already have serializer instances tied to key or values do I not. That's how I instantiate the producer/consumer.  The type is not tied to a key or value implementation only the instance which you configure the client with. I guess in theory I could share the same instance between my key and value serializers but I'm not sure I gain much there.

Reply:
> But I already have serializer instances tied to key or values do I not.

my first point is you are making the user specify this. they don't have to. 

>  I guess in theory I could share the same instance between my key and value serializers but I'm not sure I gain much there.

you gain simplicity in structure. the serializer is a simpler thing. it might seem small, but 100 like things across the library adds up to a lot.

Reply:
Realistically it's only 2 like things right. You can share the serializer the key serializer with as many producer instances as you see fit. 

I do understand what you are saying though with regard to constructing the serializer itself. We don't actually remove any inflexibility here though we just move it to another place. I would argue that most serializers will not care if they are working on a key or a value. The only reason the Avro Serializer cares is because it needs to know this information in order to register the schema using the default subject naming strategy. If it weren't for the original design of the subject registration strategy Avro too would not care if it were serializing a key or a value.

Reply:
Also note specifying a key is specific to the Avro serializer not serializers in general. 

I should actually clarify that in the doc string as the param doc is a bit misleading. I use the words `key` and `value` in the interface for a serializer but really I just mean `payload`. The fact it is the message's key or value is actually irrelevant.

Reply:
> We don't actually remove any inflexibility here though we just move it to another place.

that other place is internal, so it's hidden from the user - big difference. the way you have it results in a more complex API for no benefit.

Reply:
Honestly if I could get away with it I wouldn't even pass the topic name in as this is an example of the Registry Client API leaking into our serde API. I would prefer to just pass in a function pointer for some callable which is agnostic of both topic and key to make things incredibly simple. That's the case though and since `is_key` matches the java serializer API I'll go ahead and add it in despite my feelings about it.

Reply:
the need to pass in a `topic` 'cause of SR kind of ruins the api from the perspective of being able to beautifully specify callables inline. Furthermore, I think you'll also find that this isn't something you can do in most cases anyway. For example, how easy is it to write a bigendian/littleendian architecture aware integer serializer? I haven't looked into this in python, but i suspect it's something you wouldn't define inline (and probably something we'll provide with the library). So this idea - although very attractive (i was also attracted to it initially) - is the wrong thing to be prioritizing. what you have is kind of a middle ground that isn't optimal in any way.

given the above, I think the most important thing to be optimizing on is the point of view of someone using an existing serializer, and the callable interface should be extended to be completely general - include everything that could ever possibly be wanted (more than topic), including `is_key`. this simplifies the surface area of the parts of the library users will see/use in most cases. it's also allows for maximum flexibility. and it's also not a half-arsed middle ground - it's taking a point of view that is well justified.